### PR TITLE
[JIT] Implement Tensor.tolist()

### PR DESCRIPTION
**Pull request by SplitInfinity**
**Summary**
This commit adds an implementation of `Tensor.tolist()` to the JIT interpreter.

**Testing**
This commit adds several unit tests that test that this function works correctly for
0D, 1D, 2D and 3D tensors of type `float`, `int` and `bool`.

```
(base) meghanl-mbp:pytorch meghanl$ python test/test_jit.py TestList.test_to_list -v
Fail to import hypothesis in common_utils, tests are not derandomized
test_to_list (jit.test_list_dict.TestList)
Unit tests for Tensor.tolist() function. ... ok

----------------------------------------------------------------------
Ran 1 test in 0.329s

OK
```

### CODE REVIEW

File: aten/src/ATen/core/interned_strings.h [hunk 1]
```diff
@@ -66,6 +66,8 @@ namespace c10 {
   _(prim, StringIndex)               \
   _(prim, NumToTensor)               \
   _(prim, Uninitialized)             \
+  _(prim, ImplicitTensorToNum)       \
```
Comment (SplitInfinity) [inline aten/src/ATen/core/interned_strings.h]: Not sure how that got in here, will remove.

File: torch/csrc/jit/ir.cpp [hunk 2]
```diff
@@ -1686,6 +1686,37 @@ Value* Graph::insertUncheckedCast(Value* v, TypePtr type) {
...
+  int elem_ty = 0;
+  if (ptr == IntType::get()) {
+    elem_ty = 0;
+  } else if (ptr == FloatType::get()) {
+    elem_ty = 1;
+  } else if (ptr == BoolType::get()) {
+    elem_ty = 2;
+  } else {
+    TORCH_INTERNAL_ASSERT(false, "Unsupported data type for tolist");
+  }
```
Comment (SplitInfinity) [inline torch/csrc/jit/ir.cpp]: I don't like this either, but I have to do this to avoid capturing the output type in the operator lambda in the interpreter.

File: torch/csrc/jit/operator.cpp [hunk 3]
```diff
@@ -157,6 +157,7 @@ bool printerHasSpecialCaseFor(Symbol sym) {
       prim::CallFunction,
       prim::isinstance,
       prim::unchecked_cast,
+      prim::ToList,
```
Comment (SplitInfinity) [inline torch/csrc/jit/operator.cpp]: Should this be `aten::ToList`? Does the capitalization need to be changed?

File: torch/csrc/jit/passes/python_print.cpp [hunk 4]
```diff
@@ -1043,6 +1043,9 @@ struct PythonPrintImpl {
         }
         stmt << ")";
       } break;
+      case prim::ToList: {
+        stmt << useOf(node->input()) << ".tolist()";
+      } break;
```
Comment (SplitInfinity) [inline torch/csrc/jit/passes/python_print.cpp]: To be honest, I guessed here, but it looks like the automatic import/export tests don't run, so this isn't being tested at the moment (cc: @eellison)

**Review by SplitInfinity — COMMENTED** [review]
Just some comments/things I noticed.

File: torch/csrc/jit/passes/python_print.cpp (hunk 4 above)
Comment (eellison) [reply torch/csrc/jit/passes/python_print.cpp]: i'm trying to land https://github.com/pytorch/pytorch/pull/33468... hopefully will land soon and you can rebase on top of it.

File: torch/csrc/jit/register_prim_ops.cpp [hunk 5]
```diff
@@ -103,6 +103,70 @@ void checkImplicitTensorToNum(at::Tensor t, bool toInt) {
...
+    }
+  } else {
+    // If ty is not a list type, this invocation must be processing the last
+    // dimension, so sizes[cur_dim] consecutive scalars of the appropriate type
+    // should be read and stored in the result list.
+    if (ty == IntType::get()) {
+      for (int64_t i = 0, e = sizes[cur_dim]; i < e; ++i) {
+        int64_t scalar = *(int64_t*)data;
+        result.emplace_back(scalar);
+        data += element_size;
```
Comment (eellison) [inline torch/csrc/jit/register_prim_ops.cpp]: I'm not 100% sure but I think this should be `data += strides[dim] * elementSize`. i'm think that's what the equivalent logic in `recursive_to_list` does: https://github.com/pytorch/pytorch/blob/f531815526c69f432e46fadece44f5d3a9b70e30/torch/csrc/utils/tensor_list.cpp#L26

Also: this isn't necessary, but it might make this code a little bit easier to reason about if it had a closer correspondence to `recursive_to_list`. 

`tensorToListRecursive` could return an `IValue` instead of `c10::impl::GenericList`. When `num_tensor_dim` == `cur_dim`, return the Scalar IValue, otherwise return the GenericList.  I think this would also make it so that you no longer need to specially handle 0-dim tensors in the body of the operator.

File: torch/csrc/jit/register_prim_ops.cpp [hunk 6]
```diff
@@ -967,6 +1031,68 @@ RegisterOperators reg(
...
+             } else {
+               TORCH_INTERNAL_ASSERT(false, "Unsupported data type for tolist");
+             }
+
+             // Wrap out_ty in a ListType dim times.
+             for (int i = 0; i < dim_val; ++i) {
+               out_ty = ListType::create(out_ty);
+             }
+
+             if (auto list_type = out_ty->cast<ListType>()) {
```
Comment (eellison) [inline torch/csrc/jit/register_prim_ops.cpp]: Could you add error handling for when the tensor isn't the expected type? E.g. user said List[float] but it's a non-float tensor, or it's a tensor of the wrong dims.

File: torch/csrc/jit/script/ir_emitter.cpp [hunk 7]
```diff
@@ -2481,6 +2485,13 @@ struct to_ir {
         auto result = emitIsInstance(apply.inputs()[0], apply.inputs()[1]);
         return std::make_shared<SimpleValue>(result.value());
       }
+      case prim::ToList: {
+        auto select = Select(apply.callee());
+        auto value = select.value();
+        auto operand = emitSugaredExpr(value, 1);
```
Comment (eellison) [inline torch/csrc/jit/script/ir_emitter.cpp]: Can you add error handling for when no `type_hint` is passed in, ideally with a helpful error message ? Most users will be coming from python eager where they didn't have a type hint, so pretty much everyone who uses `tolist` will see this error message.

File: torch/csrc/jit/passes/alias_analysis.cpp [hunk 8]
```diff
@@ -371,6 +371,7 @@ void AliasDb::analyzeImpl(Node* node) {
       return;
     case prim::CallFunction:
     case prim::CallMethod:
+    case prim::ToList:
```
Comment (eellison) [inline torch/csrc/jit/passes/alias_analysis.cpp]: Since toList doesn't mutate its input and produces a fresh alias, we can put toList with the other `analyzeCreator` cases. analyzeConservative assumes that all inputs and outputs are written to.

File: test/jit/test_list_dict.py [hunk 9]
```diff
@@ -892,6 +892,99 @@ def check_list(fn, li):
...
+
+        def to_list_int_3D(x):
+            # type: (torch.Tensor) -> List[List[List[int]]]
+            li : List[List[List[int]]] = x.tolist()
+            return li
+
+        self.checkScript(to_list_int_0D, (torch.tensor(1, dtype=torch.long),))
+        int_input_1D = torch.tensor([1, 2, 3, 4], dtype=torch.long)
+        self.checkScript(to_list_int_1D, (int_input_1D,))
+        int_input_2D = torch.tensor([[1, 2, 3], [3, 4, 5]], dtype=torch.long)
```
Comment (eellison) [inline test/jit/test_list_dict.py]: Other tests to add would be no type hint, type hint doesn't match the runtime tensor type, and type hint isn't an allowed type (non-List, non-scalar).

**Review by eellison — COMMENTED** [review]
Looks great!! Adding this op cuts across a lot of different parts of the code base and different concepts so it's a tricky one to add, but this is almost ready to land.

I think there are a couple edge-cases we need to add error handling for, and we should probably wait for my PR fixing jitter tests to land before landing this.

File: torch/csrc/jit/register_prim_ops.cpp [hunk 10]
```diff
@@ -103,6 +103,70 @@ void checkImplicitTensorToNum(at::Tensor t, bool toInt) {
   }
 }
 
+// Convert the tensor pointed to by \p data to a nested list. \p dim is the
```
[the remark this answers is not in the document]
Comment (SplitInfinity) [reply torch/csrc/jit/register_prim_ops.cpp]: It's for referring to parameter in Doxygen in running text:

http://www.doxygen.nl/manual/commands.html#cmdp

File: torch/csrc/jit/passes/python_print.cpp [hunk 11]
```diff
@@ -1043,6 +1043,9 @@ struct PythonPrintImpl {
         }
         stmt << ")";
       } break;
+      case prim::ToList: {
+        stmt << useOf(node->input()) << ".tolist()";
```
[the remark this answers is not in the document]
Comment (SplitInfinity) [reply torch/csrc/jit/passes/python_print.cpp]: Looks like the land job failed :'(

File: torch/csrc/jit/operator.cpp (hunk 3 above)
Comment (SplitInfinity) [reply torch/csrc/jit/operator.cpp]: Done.

File: torch/csrc/jit/passes/python_print.cpp (hunk 4 above)
Comment (SplitInfinity) [reply torch/csrc/jit/passes/python_print.cpp]: Fixed.

File: torch/csrc/jit/ir.cpp [hunk 12]
```diff
@@ -1686,6 +1686,37 @@ Value* Graph::insertUncheckedCast(Value* v, TypePtr type) {
...
+  // Encode the base element type as an integer.
+  int elem_ty = 0;
+  if (ptr == IntType::get()) {
+    elem_ty = 0;
+  } else if (ptr == FloatType::get()) {
+    elem_ty = 1;
+  } else if (ptr == BoolType::get()) {
+    elem_ty = 2;
+  } else {
+    TORCH_INTERNAL_ASSERT(false, "Unsupported data type for tolist");
```
[the remark this answers is not in the document]
Comment (SplitInfinity) [reply torch/csrc/jit/ir.cpp]: Fixed.

File: torch/csrc/jit/passes/python_print.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (SplitInfinity) [reply torch/csrc/jit/passes/python_print.cpp]: Fixed.

File: torch/csrc/jit/register_prim_ops.cpp [hunk 13]
```diff
@@ -967,6 +1031,68 @@ RegisterOperators reg(
              return 0;
          },
          aliasAnalysisSpecialCase()),
+     Operator(
```
[the remark this answers is not in the document]
Comment (SplitInfinity) [reply torch/csrc/jit/register_prim_ops.cpp]: Fixed.

File: torch/csrc/jit/register_prim_ops.cpp [hunk 14]
```diff
@@ -103,6 +103,70 @@ void checkImplicitTensorToNum(at::Tensor t, bool toInt) {
...
+// Convert the tensor pointed to by \p data to a nested list. \p dim is the
+// number of dimensions in the tensor and \p cur_dim is the dimension being
+// processed by the current invocation. \p ty is the expected output IR type of
+// the operation. \p sizes and \p strides are the sizes and strides of the
+// tensor operand and \p element_size is the size in bytes of one tensor
+// element.
+c10::impl::GenericList tensorToListRecursive(
+    char* data,
+    int64_t cur_dim,
+    int64_t dim,
```
[the remark this answers is not in the document]
Comment (SplitInfinity) [reply torch/csrc/jit/register_prim_ops.cpp]: Fixed. Also, I don't think I used this anywhere so maybe I can just remove it altogether...

File: test/jit/test_list_dict.py [hunk 15]
```diff
@@ -892,6 +892,99 @@ def check_list(fn, li):
...
+            # type: (torch.Tensor) -> List[List[int]]
+            li : List[List[int]] = x.tolist()
+            return li
+
+        def to_list_int_3D(x):
+            # type: (torch.Tensor) -> List[List[List[int]]]
+            li : List[List[List[int]]] = x.tolist()
+            return li
+
+        self.checkScript(to_list_int_0D, (torch.tensor(1, dtype=torch.long),))
```
[the remark this answers is not in the document]
Comment (SplitInfinity) [reply test/jit/test_list_dict.py]: Done.

File: torch/csrc/jit/register_prim_ops.cpp (hunk 5 above)
Comment (SplitInfinity) [reply torch/csrc/jit/register_prim_ops.cpp]: Done and done.

I didn't realize that transposition of a tensor adjusts the strides/sizes rather than actually transposing the data. So yes, it should be `data += strides[dim] * elementSize` even if it's the last dimension. The unit tests I added for noncontiguous tensors exposed this issue. Thanks!

File: torch/csrc/jit/register_prim_ops.cpp (hunk 6 above)
Comment (SplitInfinity) [reply torch/csrc/jit/register_prim_ops.cpp]: Done.

File: torch/csrc/jit/script/ir_emitter.cpp (hunk 7 above)
Comment (SplitInfinity) [reply torch/csrc/jit/script/ir_emitter.cpp]: Done, although I'm not sure how helpful the message is? I wasn't sure how to get the expected type annotation at this point.

File: torch/csrc/jit/passes/alias_analysis.cpp (hunk 8 above)
Comment (SplitInfinity) [reply torch/csrc/jit/passes/alias_analysis.cpp]: Done.

File: test/jit/test_list_dict.py (hunk 9 above)
Comment (SplitInfinity) [reply test/jit/test_list_dict.py]: Done. I tried to add one for a tensor with an unsupported `dtype`, but I couldn't even create such a tensor lol...

File: test/jit/test_list_dict.py [hunk 16]
```diff
@@ -892,6 +892,141 @@ def check_list(fn, li):
...
+            # type: (torch.Tensor) -> List[float]
+            li : float = x.tolist()
+            return li
+
+        def to_list_unsupported_type_annotation(x):
+            # type: (torch.Tensor) -> List[float]
+            li : List[str] = x.tolist()
+            return li
+
+        def to_list_unsupported_input_type(x):
```
Comment (eellison) [inline test/jit/test_list_dict.py]: i think you forgot to compile this one

File: torch/csrc/jit/register_prim_ops.cpp [hunk 17]
```diff
@@ -967,6 +1031,56 @@ RegisterOperators reg(
...
+             } else if (elem_ty_val == 1) {
+               out_ty = FloatType::get();
+             } else if (elem_ty_val == 2) {
+               out_ty = BoolType::get();
+             } else {
+               TORCH_CHECK(false, "Unsupported element type for tolist; only int, float and bool are supported");
+             }
+
+             // Check that type of the Tensor matches that of the annotation.
+             TORCH_CHECK(tryScalarTypeFromJitType(out_ty) == t.scalar_type(), "Output annotation element type and runtime tensor element type must match for tolist()")
```
Comment (eellison) [inline torch/csrc/jit/register_prim_ops.cpp]: maybe add test for this case

File: torch/csrc/jit/register_prim_ops.cpp [hunk 18]
```diff
@@ -967,6 +1031,56 @@ RegisterOperators reg(
...
+
+             // Check that the dimension of the Tensor matches that of the annotation.
+             TORCH_CHECK(dim_val == t.dim(), "Output annotation list dimension and runtime tensor dimension must match for tolist()")
```
Comment (eellison) [inline torch/csrc/jit/register_prim_ops.cpp]: same for this case

**Review by eellison — APPROVED** [review]
Looks great, ship it !!!!

**Comment by eellison** [comment]
next time if you don't mind re-request review, i (& other people on the team) usually do our reviews off that.

File: test/jit/test_list_dict.py (hunk 16 above)
Comment (SplitInfinity) [reply test/jit/test_list_dict.py]: Meant to delete this because I cannot create a `Tensor` of a type that `tolist` does not support

File: torch/csrc/jit/register_prim_ops.cpp (hunk 18 above)
Comment (SplitInfinity) [reply torch/csrc/jit/register_prim_ops.cpp]: Done.

File: torch/csrc/jit/register_prim_ops.cpp [hunk 19]
```diff
@@ -967,6 +1034,56 @@ RegisterOperators reg(
...
+             for (int i = 0; i < dim_val; ++i) {
+               out_ty = ListType::create(out_ty);
+             }
+
+             int64_t dim = t.dim();
+             auto sizes = t.sizes();
+             auto strides = t.strides();
+             size_t element_size = t.element_size();
+             char* data = (char*)t.data_ptr();
+             auto result = tensorToListRecursive(data, 0, dim, out_ty, sizes, strides, element_size);
```
Comment (driazati) [inline torch/csrc/jit/register_prim_ops.cpp]: Is there a reason for deconstructing the tensor here instead of just passing it along?

File: torch/csrc/jit/register_prim_ops.cpp [hunk 20]
```diff
@@ -103,6 +103,73 @@ void checkImplicitTensorToNum(at::Tensor t, bool toInt) {
...
+      result.emplace_back(inner_result.toDouble());
+    } else if (inner_result.isInt()) {
+      result.emplace_back(inner_result.toInt());
+    } else if (inner_result.isBool()) {
+      result.emplace_back(inner_result.toBool());
+    } else {
+      TORCH_INTERNAL_ASSERT("Unknown return type for tensorToListRecursive");
+    }
+
+    data += strides[cur_dim] * element_size;
```
Comment (driazati) [inline torch/csrc/jit/register_prim_ops.cpp]: I'm not sure if it will work here but using `TensorIterator` will handle things like strides for you, instead of manually looping over the tensor's data

File: torch/csrc/jit/register_prim_ops.cpp (hunk 19 above)
Comment (SplitInfinity) [reply torch/csrc/jit/register_prim_ops.cpp]: I basically copied the Python implementation and that's what that does.

File: torch/csrc/jit/register_prim_ops.cpp (hunk 20 above)
Comment (SplitInfinity) [reply torch/csrc/jit/register_prim_ops.cpp]: I'm new here and I basically copied the Python implementation. If you feel strongly about this I can do it in a follow up PR.

File: torch/csrc/jit/register_prim_ops.cpp [hunk 21]
```diff
@@ -967,6 +1034,56 @@ RegisterOperators reg(
...
+             // Check that the dimension of the Tensor matches that of the annotation.
+             TORCH_CHECK(dim_val == t.dim(), "Output annotation list dimension and runtime tensor dimension must match for tolist()")
+
+             // Wrap out_ty in a ListType dim times.
+             for (int i = 0; i < dim_val; ++i) {
+               out_ty = ListType::create(out_ty);
+             }
+
+             int64_t dim = t.dim();
+             auto sizes = t.sizes();
```
Comment (eellison) [inline torch/csrc/jit/register_prim_ops.cpp]: hey sorry, one last thing, we need to move it to CPU if it's not already 
https://github.com/pytorch/pytorch/blob/f531815526c69f432e46fadece44f5d3a9b70e30/torch/csrc/utils/tensor_list.cpp#L33

**Review by eellison — COMMENTED** [review]
one last thing we forgot. i think the close equivalence to the python implementation is a good thing because it makes it easier to review / maintain. i did that with torch.tensor (equal and opposite op) and we haven't had any bugs from it.

File: torch/csrc/jit/register_prim_ops.cpp (hunk 21 above)
Comment (eellison) [reply torch/csrc/jit/register_prim_ops.cpp]: you can do this in a follow up as well
